mPower2/mPower2/PassiveCollectors.swift

```diff
@@ -0,0 +1,291 @@
+        // even if authorization is currently not adequate for our needs, if that changes we'll get
+        // a callback letting us know, and we can start monitoring then.
+        locationManager = CLLocationManager()
+        locationManager!.delegate = self
+        locationManager!.allowsBackgroundLocationUpdates = true
+        locationManager!.pausesLocationUpdatesAutomatically = false
+        
+        func authorizationError(for status: CLAuthorizationStatus) -> Error {
+            let rsd_status: RSDAuthorizationStatus = (status == .restricted) ? .restricted : .denied
+            let error = RSDPermissionError.notAuthorized(.location, rsd_status)
+            #if DEBUG
+            print("Not authorized to start the passive location recorder: \(rsd_status)")
+            #endif
+            return error
+        }
+        
+        // Get the current status and exit early if the status is restricted or denied.
+        let status = CLLocationManager.authorizationStatus()
+        if status != .authorizedAlways && status != .notDetermined {
+            // anything but "always" or "not yet asked" means we can't continue
+            throw authorizationError(for: status)
+        }
+        
+        guard status == .notDetermined else { return }
+
+        // Authorization had not yet been given or denied, so we need to do that now and check again.
+        locationManager!.requestAlwaysAuthorization()
+        
+        let newStatus = CLLocationManager.authorizationStatus()
+        if newStatus != .authorizedAlways {
+            // and here we are again
+            throw authorizationError(for: newStatus)
+        }
+    }
+
+    /// Start the passive displacement collector.
+    func start() {
+        // ...if we can.
+        guard CLLocationManager.significantLocationChangeMonitoringAvailable() else { return }
+        
+        DispatchQueue.main.async {
+            if self.locationManager == nil {
+                do {
+                    try self.setupLocationManager()
+                } catch {
+                    // we don't have permission to track their location passively so just ignore it
+                    return
+                }
+            }
+
+            self.locationManager!.startMonitoringSignificantLocationChanges()
+        }
+    }
+    
+    /// Stop the passive displacement collector.
+    /// - parameter discarding: Ignored by this collector; it gets data infrequently enough to just upload as it goes.
+    func stop(discarding: Bool) {
+        // if there's no locationManager, there's nothing to do
+        guard locationManager != nil else { return }
+        
+        self.locationManager!.stopMonitoringSignificantLocationChanges()
+    }
+    
+    /// Given a current and previous location reading, generate a displacement data point.
+    func displacementData(for location: CLLocation, previousLocation: CLLocation) -> [String: RSDJSONValue] {
```

Comment:
In fact, if you add `floor` to the `RSDDistanceRecord` then you could use it directly.

```
    /// Given a current and previous location reading, generate a displacement data point.
    func displacementData(for location: CLLocation, previousLocation: CLLocation) -> RSDDistanceRecord {
        return RSDDistanceRecord(uptime: ProcessInfo.processInfo.systemUptime,
                                 timestamp: 0,
                                 stepPath: "Passive",
                                 location: location,
                                 previousLocation: previousLocation,
                                 totalDistance: nil)
    }
    
    func uploadDisplacement(_ displacementData: RSDDistanceRecord) {
        let archiveFilename = kPassiveDisplacementArchiveFilename
        let archive = SBBDataArchive(reference: schemaIdentifier, jsonValidationMapping: nil)
        if let schemaRevision = MP2BridgeConfiguration.shared.schemaInfo(for: schemaIdentifier)?.schemaVersion {
            archive.setArchiveInfoObject(schemaRevision, forKey: kSchemaRevisionKey)
        }
        
        do {
            let data = try displacementData.rsd_jsonEncodedData()
            archive.insertData(intoArchive: data, filename: archiveFilename, createdOn: Date())
            try archive.complete()
            archive.encryptAndUploadArchive()
        }
        catch let err {
            debugPrint("Failed to encode displacement data. \(err)")
        }
   }
```

Reply:
I mean, I can sort of see this, I actually considered trying to reuse RSDDistanceRecord(er) in some way, but it's really for a totally different purpose. We're not starting and stopping a recorder for a specific activity, and we're not trying to accurately calculate the distance they travel during that activity. We're just trying to get a general idea of how much they're getting out and about. The uptime and timestamp fields don't really make sense for this, course will almost never be available and isn't the same as the bearing from the previous location, and we don't really want or need the app deciding when to start and stop accumulating relative distances. Instead I ended up adapting the fields from mPower 1, discarding the "unit" fields because they're all always SI units (meters, meters/second), and uploading them as individual rows in a table rather than collecting a days' worth into a CSV file which the researchers were not fond of.

Reply:
Looking more closely, never mind--you're right, this does make sense. I'm going to just add timestampUnix, floor, and bearingRadians fields to RSDDistanceRecord and use that.